File: server/src/routes/room.ts
import { Router, Request, Response } from 'express';
import { createRoom, getRoomByName, verifyRoomPassword, getRoomById } from '../db/queries/room';
import { addMemberToRoom, removeMemberFromRoom, getRoomMember } from '../db/queries/roomMember';
import { getUserById } from '../db/queries/user';
import sequelize from '../db/connection';
import logger from '../config/logger';

const router = Router();


router.post('/create', async (req: Request, res: Response): Promise<void> => {
  try {
    const { name, password } = req.body;
    
    // check if room name already exists
    const existingRoom = await getRoomByName(name);
    if (existingRoom) {
      res.status(400).json({ error: 'Room name already exists' });
      return;
    }

    // create new room
    const room = await createRoom(name, password);
    
    res.json({
      id: room.id,
      name: room.name,
      createdTime: room.createdTime
    });
  } catch (error) {
    logger.error('Failed to create room:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.get('/query', async (req: Request, res: Response): Promise<void> => {
  const { name } = req.query;
  try {
    const room = await getRoomByName(name as string);
    if (!room) {
    res.status(404).json({ error: 'Room not found' });
    return;
  }
    res.json(room);
  } catch (error) {
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.post('/join', async (req: Request, res: Response): Promise<void> => {
  const transaction = await sequelize.transaction();
  
  try {
    const { roomId, userId, password } = req.body;
    
    // check if user exists
    const user = await getUserById(userId);
    if (!user) {
      res.status(404).json({ error: 'User not found' });
      return;
    }

    // check if room exists
    const room = await getRoomById(roomId);
    if (!room) {
      res.status(404).json({ error: 'Room not found' });
      return;
    }

    // check if password is valid
    const isPasswordValid = await verifyRoomPassword(room, password || '');
    if (!isPasswordValid) {
      res.status(401).json({ error: 'Invalid password' });
      return;
    }

    // check if user is already in the room
    const existingMember = await getRoomMember(roomId, userId);
    if (existingMember) {
      res.status(400).json({ error: 'User is already in the room' });
      return;
    }

    // add user to room
    const member = await addMemberToRoom(roomId, userId, false, false, transaction);
    
    await transaction.commit();
    // TODO: set cookie
    res.json({
      roomId: member.roomId,
      userId: member.userId,
      isAdmin: member.isAdmin,
      canGrantAdmin: member.canGrantAdmin
    });
  } catch (error) {
    await transaction.rollback();
    logger.error('Failed to join room:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

router.post('/leave', async (req: Request, res: Response): Promise<void> => {
  try {
    const { roomId, userId } = req.body;

    // check if user is in the room
    const member = await getRoomMember(roomId, userId);
    if (!member) {
      res.status(404).json({ error: 'User is not in the room' });
      return;
    }

    // remove user from room
    try {
      await removeMemberFromRoom(roomId, userId);
    } catch (error) {
      logger.error('Failed to remove member from room:', error);
      res.status(500).json({ error: 'Internal server error' });
      return;
    }
    
    res.json({ message: 'Successfully left the room' });
  } catch (error) {
    logger.error('Failed to leave room:', error);
    res.status(500).json({ error: 'Internal server error' });
  }
});

export default router; 

File: server/internal/handlers/room.go
package handlers

import (
	"fmt"
	"net/http"
	"sync-player-server/internal/config"
	"sync-player-server/internal/database"
	"sync-player-server/internal/middleware"
	"sync-player-server/internal/utils"

	"github.com/gin-gonic/gin"
	"gorm.io/gorm"
)

// RoomCreate handles room creation
func RoomCreate(c *gin.Context) {
	var req struct {
		Name     string `json:"name" binding:"required"`
		Password string `json:"password"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	existingRoom, err := database.GetRoomByName(req.Name)
	if err == nil && existingRoom != nil {
		c.JSON(http.StatusOK, gin.H{
			"id":          existingRoom.ID,
			"name":        existingRoom.Name,
			"createdTime": existingRoom.CreatedTime,
		})
		return
	}

	var password *string
	if req.Password != "" {
		password = &req.Password
	}

	room, err := database.CreateRoom(req.Name, password)
	if err != nil {
		config.Logger.Errorf("Failed to create room: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	c.JSON(http.StatusOK, gin.H{
		"id":          room.ID,
		"name":        room.Name,
		"createdTime": room.CreatedTime,
	})
}

// RoomQuery handles room query by name
func RoomQuery(c *gin.Context) {
	name := c.Query("name")
	if name == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Room name is required"})
		return
	}

	room, err := database.GetRoomByName(name)
	if err != nil {
		if err == gorm.ErrRecordNotFound {
			c.JSON(http.StatusNotFound, gin.H{"error": "Room not found"})
			return
		}
		config.Logger.Errorf("Failed to query room: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	c.JSON(http.StatusOK, room)
}

// RoomJoin handles user joining a room
func RoomJoin(c *gin.Context) {
	var req struct {
		RoomID   uint   `json:"roomId" binding:"required"`
		UserID   uint   `json:"userId" binding:"required"`
		Password string `json:"password"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	err := database.DB.Transaction(func(tx *gorm.DB) error {
		user, err := database.GetUserByID(req.UserID)
		if err != nil {
			c.JSON(http.StatusNotFound, gin.H{"error": "User not found"})
			return err
		}

		room, err := database.GetRoomByID(req.RoomID)
		if err != nil {
			c.JSON(http.StatusNotFound, gin.H{"error": "Room not found"})
			return err
		}

		if !database.VerifyRoomPassword(room, req.Password) {
			c.JSON(http.StatusUnauthorized, gin.H{"error": "Invalid password"})
			return gorm.ErrInvalidData
		}

		existingMember, _ := database.GetRoomMember(req.RoomID, req.UserID)
		if existingMember != nil {
			c.JSON(http.StatusBadRequest, gin.H{"error": "User is already in the room"})
			return gorm.ErrDuplicatedKey
		}

		newMember, err := database.AddMemberToRoom(req.RoomID, req.UserID, false, false, tx)
		if err != nil {
			config.Logger.Errorf("Failed to add member to room: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
			return err
		}

		// Generate JWT token
		token, err := utils.GenerateJWT(user.ID, user.Username, req.RoomID, newMember.IsAdmin)
		if err != nil {
			config.Logger.Errorf("Failed to generate JWT token: %v", err)
			c.JSON(http.StatusInternalServerError, gin.H{"error": "Failed to generate authentication token"})
			return err
		}

		// Also set cookie for backward compatibility
		middleware.SetUserInfoCookie(c, req.RoomID, user.ID)

		c.JSON(http.StatusOK, gin.H{
			"token":         token,
			"roomId":        newMember.RoomID,
			"userId":        newMember.UserID,
			"isAdmin":       newMember.IsAdmin,
			"canGrantAdmin": newMember.CanGrantAdmin,
		})

		return nil
	})

	if err != nil {
		return
	}
}

// RoomLeave handles user leaving a room
func RoomLeave(c *gin.Context) {
	var req struct {
		RoomID uint `json:"roomId" binding:"required"`
		UserID uint `json:"userId" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request body"})
		return
	}

	_, err := database.GetRoomMember(req.RoomID, req.UserID)
	if err != nil {
		c.JSON(http.StatusNotFound, gin.H{"error": "User is not in the room"})
		return
	}

	if err := database.RemoveMemberFromRoom(req.RoomID, req.UserID); err != nil {
		config.Logger.Errorf("Failed to remove member from room: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	c.JSON(http.StatusOK, gin.H{"message": "Successfully left the room"})
}

// RoomQueryOnlineUsers handles querying online users in a room
func RoomQueryOnlineUsers(c *gin.Context) {
	roomIDStr := c.Query("roomId")
	if roomIDStr == "" {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Room ID is required"})
		return
	}

	var roomID uint
	if _, err := fmt.Sscanf(roomIDStr, "%d", &roomID); err != nil {
		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid room ID"})
		return
	}

	users, err := database.GetOnlineUsers(roomID)
	if err != nil {
		config.Logger.Errorf("Failed to query online users: %v", err)
		c.JSON(http.StatusInternalServerError, gin.H{"error": "Internal server error"})
		return
	}

	c.JSON(http.StatusOK, users)
}
